File: 프로그래머스/2/49994. 방문 길이/방문 길이.py
def isValid(nx, ny):
    return 0 <= nx < 11 and 0 <= ny < 11

def nextLoc(x, y, direc):
    if direc == "U": nx, ny = x, y + 1
    elif direc == "D": nx, ny = x, y - 1
    elif direc == "L": nx, ny = x + 1, y
    elif direc == "R": nx, ny = x - 1, y
    return nx, ny

def solution(dirs):
    answer = set()
    x,y = 5,5
    
    for direc in dirs:
        nx, ny = nextLoc(x, y, direc)
        if not isValid(nx, ny): continue
        answer.add((x,y,nx,ny))
        answer.add((nx,ny,x,y))
        x, y = nx, ny
        
    return len(answer) / 2

File: 프로그래머스/2/49994. 방문 길이/방문 길이.js
function solution(dirs) {
    let moves = {L: [-1, 0], R: [1, 0], U: [0, 1], D: [0, -1]};
    let s = [0, 0];
    let route = new Set();
    
    for (let dir of dirs) {
        let nx = s[0] + moves[dir][0];
        let ny = s[1] + moves[dir][1];
        
        if (nx > 5 || nx < -5 || ny > 5 || ny < -5) continue;
        
        route.add("" + s[0] + s[1] + nx + ny);
        route.add("" + nx + ny + s[0] + s[1]);
        
        s = [nx, ny];
    }
    
    return route.size / 2;
}